### ISSUE TITLE
Error in temperature and brightness when using lightOperation

### ISSUE THREAD

**Issue by freahs** [AlCalzone/node-tradfri-client#15]
Only tested on "TRADFRI bulb E27 WS opal 980lm" with firmware 1.2.217 (Gateway firmware is 1.2.42).

To reproduce, first set the brightness of the light with

```javascript
client.operateLight(light, {"dimmer":3})
```

When inspecting `light` it will have its `dimmer` property set to 4. Likewise, 
```javascript
client.operateLight(light, {"colorTemperature":100})
```
will result in a `colorTemperature` of 98 in the light object.

There are several more values that will not get set correctly in similar fashion. I suppose this is due to conversion between values used by the gateway and the [0,100] scale used in the API.

It would be great if the proper ranges for both brightness and temperature were usable in the API, since it allows finer granularity control. You probably already know this but:
   - brightness (property 5851) accepts values in [0,254]
   - color temperature (property 5711) accepts any value but WS bulbs will show values in [250, 454]. CWS bulbs supposedly can't use this property but I wouldn't know since I don't own any.

**Comment by AlCalzone** [comment]
I'll have to look into it. Its very likely due to internal rounding. The percentage scale is a concious decision to make the lib easier to use, but I get your point regarding the granularity. Maybe I'll add another parameter to let the user decide to use the raw values.

**Comment by freahs** [comment]
Great, thanks!

**Comment by AlCalzone** [comment]
A couple of updates:
* The brightness rounding should be fixed in e920c64.
* The color temperature is currently set through the colorX/colorY values. In my tests, the gateway would only accept 3 distinct values for the colorTemperature property (250, 370 and 454) while colorX/Y supports the whole spectrum. I'm working on fixing that bug.
* I'll add support for using floats internally, so the rounding only has to happen between `node-tradfri-client` and user code.
* Additionally, I'll add a flag telling the client to use the raw CoAP values instead of the normalized ranges. But I'm not sure yet how to best do that, as it has to be defined per `TradfriClient` instance and be accessible in the modules.

**Comment by AlCalzone** [comment]
Looking at the conversion functions, I see no reason that setting a `colorTemperature` of 100 should result in a value of 98 other than the gateway reporting a different value. Could you test the following please?
1. Define a custom logger function for the tradfri client:
```js
function log(msg, sev) {
    if (sev === "debug") console.log(msg); // or some other output
}
```

2. Pass that function to the client constructor:
```js
const tradfri = new TradfriClient(
    hostname, log
);
```

3. Change the color temperature and send me the log output for each call:
```js
client.operateLight(light, {"colorTemperature":100}); // and some other values if possible
```

**Comment by freahs** [comment]
I set the light `colorTemperature` to 50 and ran `client.operateLight(light, {"colorTemperature":0, "transitionTime":0});` three times, one second apart, which gave the following log output:

> updateResource(15001/65537) > sending payload: {"3311":[{"5709":24930,"5710":27000,"5712":0}]}
> observeDevice > {"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb E27 WS opal 980lm","2":"","3":"1.2.217","6":1},"3311":[{"5706":"0","5709":24930,"5710":27000,"5711":0,"5717":0,"5850":1,"5851":254,"9003":0}],"5750":2,"9001":"tak","9002":1504805791,"9003":65537,"9019":1,"9020":1515843951,"9054":0}
> observeDevice > {"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb E27 WS opal 980lm","2":"","3":"1.2.217","6":1},"3311":[{"5706":"f5faf6","5709":24841,"5710":24593,"5711":250,"5717":0,"5850":1,"5851":254,"9003":0}],"5750":2,"9001":"tak","9002":1504805791,"9003":65537,"9019":1,"9020":1515843951,"9054":0}

The second time:

> updateResource(15001/65537) > sending payload: {"3311":[{"5709":24930,"5710":27000,"5712":0}]}
> observeDevice > {"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb E27 WS opal 980lm","2":"","3":"1.2.217","6":1},"3311":[{"5706":"0","5709":24930,"5710":27000,"5711":0,"5717":0,"5850":1,"5851":254,"9003":0}],"5750":2,"9001":"tak","9002":1504805791,"9003":65537,"9019":1,"9020":1515843951,"9054":0}

And then third time: 

> updateResource(15001/65537) > empty object, not sending any payload

The callback for `'device updated'` are called each time observeDevice are logged. In all the above cases, `colorTemperature` was 0 in the `Accessory` sent to the callback.

Then I ran `client.operateLight(light, {"colorTemperature":100, "transitionTime":0});` three times, one second apart:

> updateResource(15001/65537) > sending payload: {"3311":[{"5709":33135,"5710":27000,"5712":0}]}
> observeDevice > {"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb E27 WS opal 980lm","2":"","3":"1.2.217","6":1},"3311":[{"5706":"0","5709":33135,"5710":27000,"5711":0,"5717":0,"5850":1,"5851":254,"9003":0}],"5750":2,"9001":"tak","9002":1504805791,"9003":65537,"9019":1,"9020":1515843951,"9054":0}
> observeDevice > {"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb E27 WS opal 980lm","2":"","3":"1.2.217","6":1},"3311":[{"5706":"efd275","5709":32977,"5710":27105,"5711":454,"5717":0,"5850":1,"5851":254,"9003":0}],"5750":2,"9001":"tak","9002":1504805791,"9003":65537,"9019":1,"9020":1515843951,"9054":0}

In the above, the first time the callback is called, `colorTemperature` are 100 in the `Accessory`, but the second time it's 98.

The second time:

> updateResource(15001/65537) > sending payload: {"3311":[{"5709":33135,"5710":27000,"5712":0}]}
> observeDevice > {"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb E27 WS opal 980lm","2":"","3":"1.2.217","6":1},"3311":[{"5706":"0","5709":33135,"5710":27000,"5711":0,"5717":0,"5850":1,"5851":254,"9003":0}],"5750":2,"9001":"tak","9002":1504805791,"9003":65537,"9019":1,"9020":1515843951,"9054":0}

This time the `colorTemperature` are 100. 

Then third time:

> updateResource(15001/65537) > empty object, not sending any payload

As for the color temperature, it can be set with the property 5711. Try the following which works for white spectrum bulbs:

```javascript
let instanceId = 12345 // some proper id of a white spectrum bulb.
for (let t = 250; t <= 454; t++) {
	await client.request('15001/' + instanceId, 'put', {3311:[{5711:t}]});

	// too many concecutive updates will make the gateway unstable
	await (async () => {
		return new Promise(r => setTimeout(r, 200));.
	})();
}
```

**Comment by AlCalzone** [comment]
Thanks. On first glance it looks like the gateway tries once to remap the given values to the predefined ones. But those look different - they used to be [24930..33135], now they seem to be [24841..32977].

I'll have to try with your test again, last time I checked, the gateway did not accept any values for property 5711 other than 250, 370 and 454.

**Comment by AlCalzone** [comment]
Nice, that actually works now. I'm going to drop the colorTemp => colorX crutch immediately

**Comment by AlCalzone** [comment]
Both issues should be fixed in v0.6.0. Please test!

**Comment by AlCalzone** [comment]
Using raw coap values ([0..254] etc) is supported in v0.7.0. Refer to the docs on how to enable and use it.
